src/redis/cache_impl.go

```diff
@@ -222,6 +249,12 @@ func (this *rateLimitCacheImpl) DoLimit(
 				// in the near limit range.
 				limits[i].Stats.NearLimit.Add(uint64(overLimitThreshold - max(nearLimitThreshold, limitBeforeIncrease)))
 			}
+			if this.localCache != nil {
```

Comment:
One thing to point out here is for longer limits, this might effectively double the time that something is rate limited. Meaning, let's say you have an hour limit. If you get over limit at the end of the hour and are about to reset, you would have this entry in the cache for the entire duration of the limit, right?

One option would be to set the cache TTL to only the number of seconds remaining in the limit, since you know the current time and the divider. The other option for now would be to clearly document this case with comments and in the end user docs.

Reply:
I just had a similar conversation with Henry yesterday. I think it doesn't matter whether setting to be the remaining duration or the entire duration. Based on current design, both the redis cache key and local_cache key are similar to `actualKey_duration`. For example, if we have an hour limit on all mongo connections, the cache key would be similar to `mongo_1h`, `mongo_2h`.  let's say, in the hour 1 (0h0m - 0h59m), we started to get ratelimited in the 50th minutes, and the ttl of local_cache will be set as 1 hour(0h50m-1h49m), the cache key is `mongo_1h`. In the time of 1h1m, since either the local cache key or redis cache key becomes different (mongo_2h), it won't get ratelimited.
Here is how cache key gets generated: https://github.com/lyft/ratelimit/blob/8d923502fb14fa2b2d31c945676dcea503f11e0b/src/redis/cache_impl.go#L74-L82

Since it requires reading the redis cache to calculate the remaining duration, it should be better to set the entire duration.

I'll add the comment to make it clear.

Reply:
In this simpler approach, we store 2x the number of overlimited keys.  The average case is that we have 100s of overlimited rules, but in the worst case the memory bounded LRU cache will evict the old key we no longer need.

Reply:
I see OK, that makes sense that the cache key changes. Thanks for the explanation. Yeah mor comments would be good as well as the end-user docs.

Reply:
just add comments about it.